From 61bd3da80eb0f9803d4b935fbbfea676986a24a6 Mon Sep 17 00:00:00 2001
From: Vijay Balaji <quic_vijbal@quicinc.com>
Date: Tue, 4 Jun 2024 16:57:19 +0530
Subject: [PATCH] sysupgrade : fix sysupgrade with alive upgrade change

This change adds support to alive upgrade in sysupgrade
feature. alive upgrade can be done with '-a' option.
With this change, able to detach ubi to upgrade with ubi
image in respective partition(ubi0).

Change-Id: Idd3b7684f2d223079c2817c6aea377f92b68122d
Signed-off-by: Vijay Balaji <quic_vijbal@quicinc.com>
(cherry picked from commit 3bde4eb602cd24a6242f7f5cf2bd61773d27ba3a)
---
 package/base-files/files/sbin/sysupgrade | 47 +++++++++++++++---------
 1 file changed, 30 insertions(+), 17 deletions(-)

diff --git a/package/base-files/files/sbin/sysupgrade b/package/base-files/files/sbin/sysupgrade
index 611d8830bfd1..4773d6f2d52a 100755
--- a/package/base-files/files/sbin/sysupgrade
+++ b/package/base-files/files/sbin/sysupgrade
@@ -32,6 +32,7 @@ export VERBOSE=1
 export SAVE_CONFIG=1
 export IGNORE_MINOR_COMPAT=0
 export FORCE=0
+export ALIVE=0
 export CONFFILES=/tmp/sysupgrade.conffiles
 
 # parse options
@@ -46,6 +47,7 @@ while [ -n "$1" ]; do
 		-p) SAVE_PARTITIONS=0;;
 		-k) SAVE_INSTALLED_PKGS=1;;
 		-u) SKIP_UNCHANGED=1;;
+		-a|--alive) export ALIVE=1;;
 		-b|--create-backup) CONF_BACKUP="$2" NEED_IMAGE=1; shift;;
 		-r|--restore-backup) CONF_RESTORE="$2" NEED_IMAGE=1; shift;;
 		-l|--list-backup) CONF_BACKUP_LIST=1;;
@@ -63,6 +65,9 @@ while [ -n "$1" ]; do
 	shift;
 done
 
+export ARGV="$*"
+export ARGC="$#"
+
 print_help() {
 	cat <<EOF
 Usage: $0 [<upgrade-option>...] <image file or URL>
@@ -75,6 +80,7 @@ upgrade-option:
 	-o           attempt to preserve all changed files in /, except those
 	             from packages but including changed confs.
 	-u           skip from backup files that are equal to those in /rom
+	-a           alive upgrade valid only when trymode is enabled
 	-n           do not save configuration over reflash
 	-p           do not attempt to restore the partition table after flash.
 	-k           include in backup a list of current installed packages at
@@ -336,6 +342,8 @@ if [ -n "$CONF_RESTORE" ]; then
 	exit $?
 fi
 
+echo $ALIVE > /tmp/.alive_upgrade
+
 type platform_check_image >/dev/null 2>/dev/null || {
 	echo "Firmware upgrade is not implemented for this platform." >&2
 	exit 1
@@ -401,22 +409,27 @@ if [ $TEST -eq 1 ]; then
 	exit 0
 fi
 
-install_bin /sbin/upgraded
-v "Commencing upgrade. Closing all shell sessions."
-
-if [ -n "$FAILSAFE" ]; then
-	printf '%s\x00%s\x00%s' "$RAM_ROOT" "$IMAGE" "$COMMAND" >/tmp/sysupgrade
-	lock -u /tmp/.failsafe
+if [ $ALIVE -eq 1 ]; then
+	do_upgrade
 else
-	json_init
-	json_add_string prefix "$RAM_ROOT"
-	json_add_string path "$IMAGE"
-	[ $FORCE -eq 1 ] && json_add_boolean force 1
-	[ $SAVE_CONFIG -eq 1 ] && json_add_string backup "$CONF_TAR"
-	json_add_string command "$COMMAND"
-	json_add_object options
-	json_add_int save_partitions "$SAVE_PARTITIONS"
-	json_close_object
-
-	ubus call system sysupgrade "$(json_dump)"
+
+	install_bin /sbin/upgraded
+	v "Commencing upgrade. Closing all shell sessions."
+
+	if [ -n "$FAILSAFE" ]; then
+		printf '%s\x00%s\x00%s' "$RAM_ROOT" "$IMAGE" "$COMMAND" >/tmp/sysupgrade
+		lock -u /tmp/.failsafe
+	else
+		json_init
+		json_add_string prefix "$RAM_ROOT"
+		json_add_string path "$IMAGE"
+		[ $FORCE -eq 1 ] && json_add_boolean force 1
+		[ $SAVE_CONFIG -eq 1 ] && json_add_string backup "$CONF_TAR"
+		json_add_string command "$COMMAND"
+		json_add_object options
+		json_add_int save_partitions "$SAVE_PARTITIONS"
+		json_close_object
+
+		ubus call system sysupgrade "$(json_dump)"
+	fi
 fi
-- 
2.34.1

